exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionImplementationRegistry.java

```diff
@@ -186,4 +226,105 @@ public boolean isFunctionComplexOutput(String name) {
+   */
+  public boolean loadRemoteFunctions() {
+    List<String> missingJars = Lists.newArrayList();
+    Registry registry = remoteFunctionRegistry.getRegistry();
+
+    List<String> localJars = drillFuncRegistry.getAllJarNames();
+    for (Jar jar : registry.getJarList()) {
+      if (!localJars.contains(jar.getName())) {
+        missingJars.add(jar.getName());
+      }
+    }
+
+    for (String jarName : missingJars) {
+      try {
+        Path localUdfArea = new Path(new File(getUdfDir()).toURI());
+        Path registryArea = remoteFunctionRegistry.getRegistryArea();
+        FileSystem fs = remoteFunctionRegistry.getFs();
+
+        String sourceName = JarUtil.getSourceName(jarName);
+
+        Path remoteBinary = new Path(registryArea, jarName);
+        Path remoteSource = new Path(registryArea, sourceName);
+
+        Path binary = new Path(localUdfArea, jarName);
+        Path source = new Path(localUdfArea, sourceName);
+
+        fs.copyToLocalFile(remoteBinary, binary);
+        fs.copyToLocalFile(remoteSource, source);
+
+        URL[] urls = {binary.toUri().toURL(), source.toUri().toURL()};
+        ClassLoader classLoader = new URLClassLoader(urls);
+        ScanResult scanResult = scan(classLoader, binary, urls);
+        drillFuncRegistry.register(jarName, scanResult, classLoader);
+      } catch (IOException | FunctionValidationException e) {
+        logger.error("Problem during remote functions load from {}", jarName, e);
+      }
+    }
+
+    return missingJars.size() > 0;
+  }
+
+  private ScanResult scan(ClassLoader classLoader, Path path, URL[] urls) throws IOException {
+    Enumeration<URL> e = classLoader.getResources(CommonConstants.DRILL_JAR_MARKER_FILE_RESOURCE_PATHNAME);
+    while (e.hasMoreElements()) {
+      URL res = e.nextElement();
+      if (res.getPath().contains(path.toUri().getPath())) {
+        DrillConfig drillConfig = DrillConfig.create(ConfigFactory.parseURL(res));
+        return RunTimeScan.dynamicPackageScan(drillConfig, Sets.newHashSet(urls));
+      }
+    }
+    throw new FunctionValidationException(String.format("Marker file %s is missing in %s.",
+        CommonConstants.DRILL_JAR_MARKER_FILE_RESOURCE_PATHNAME, path.getName()));
+  }
+
+  private static String getUdfDir() {
+    return Preconditions.checkNotNull(System.getenv("DRILL_UDF_DIR"), "DRILL_UDF_DIR variable is not set");
+  }
+
+  /**
+   * Fires when jar name is submitted for unregistration.
+   * Will unregister all functions associated with the jar name
+   * and delete binary and source associated with the jar from local DRILL_UDF_DIR.
+   */
+  public class UnregistrationListener implements TransientStoreListener {
+
+    @Override
+    public void onChange(TransientStoreEvent event) {
+        String jarName = (String) event.getValue();
+        String sourceName = JarUtil.getSourceName(jarName);
+        String localDir = getUdfDir();
+        FileUtils.deleteQuietly(new File(localDir, jarName));
```

Comment:
Race condition? We delete quietly. How will this work if another thread is using the function or jar? Is the corresponding class loader removed? Or, do we allow the class loader (and its loaded classes) to live forever? Maybe want to discuss the clean-up process in some detail in a comment.